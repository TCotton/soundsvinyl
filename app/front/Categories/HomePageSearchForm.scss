@import '../../assets/sass/tools';

.search {
	.searchForm {
		input[type=search] {
			appearance: none;
			-webkit-appearance: textfield;
			min-width: 75%;
			height: 30px;
			line-height: 30px;
			padding: 5px;
			border: 1px solid $footerBackground;
			box-shadow: inset 0 5p 5px darken($footerBackground, 50%);
			margin-bottom: 50px;
		}
	}

	.searchName {
		> p {
			@include font-calculator($font_family_header, 13px, 1);
			color: $secondHeaderColour;
			font-weight: 300;
			line-height: 1.5em;
			margin-top: 20px;
		}
	}

	h4 {
		@include font-calculator($font_family_header, 16px, 0.5);
		color: $secondHeaderColour;
		font-weight: 300;
		line-height: 1.1em;
	}
}
