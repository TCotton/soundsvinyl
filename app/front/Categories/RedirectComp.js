import React, { Component } from 'react';
import './categoriesHomepage.scss';
import PropTypes from 'prop-types';
import { Redirect, withRouter, Switch } from 'react-router-dom';

class RedirectComp extends Component {

	static propTypes = {
		search: PropTypes.oneOfType([
			PropTypes.string,
			PropTypes.bool
		])
	}

	static defaultProps = {
		search: false
	}

	constructor ( props ) {
		super( props );
	}

	render () {
		const { search } = this.props;

		if( !search ) {
			return null;
		}

		if( search ) {
			return (
				<Switch>
					<Redirect
						push
						to={search}
					/>
				</Switch>
			)
		}
	}
}

export default withRouter( RedirectComp );
