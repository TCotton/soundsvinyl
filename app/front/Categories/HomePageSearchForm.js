import React, { Component } from 'react';
import './HomePageSearchForm.scss';
import { connect } from 'react-redux';
import RedirectComp from './RedirectComp';
import { withRouter } from 'react-router-dom'

export class HomePageSearchForm extends Component {

	constructor ( props ) {
		super( props );

		this.state = {
			error: null,
			message: false,
			search: '',
			redirectRoute: false
		};

		this.handleSubmit = this.handleSubmit.bind( this );
		this.handleInputChange = this.handleInputChange.bind( this );
	}

	handleSubmit ( event ) {
		event.preventDefault();
		const { search } = this.state;

		this.setState( {
			redirectRoute: `/category/${search}`
		} );
	}

	handleInputChange ( event ) {

		const target = event.target;
		const value = target.value;
		const name = target.name;

		this.setState( {
			[ name ]: value
		} );

	}

	render () {
		const { search, error, message, redirectRoute } = this.state;

		return (
			<React.Fragment>
				<RedirectComp
					search={redirectRoute}
				/>
				<div styleName='search'>
					<h4>
						Search by category
					</h4>
					<form
						onSubmit={this.handleSubmit}
						styleName='searchForm'
					>
						<label
							className='visuallyhidden'
							htmlFor='search'
						>
							Your search term
						</label>
						<input
							data-autocapitalize='none'
							id='search'
							name='search'
							onChange={this.handleInputChange}
							type='search'
							value={search}
						/>
					</form>
					<div styleName='searchName'>
						{error}
						{message &&
						<p>
							There are no results for that search term.
							<br />
							Try another term.
						</p>}
					</div>
				</div>
			</React.Fragment>
		)
	}
}

export default connect()( withRouter( HomePageSearchForm ) );

