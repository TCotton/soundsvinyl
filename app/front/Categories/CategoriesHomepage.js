import React, { Component } from 'react'
import './CategoriesHomepage.scss'
import PageUnit from '../Page/PageUnit'
import HomePageSearchForm from './HomePageSearchForm'
import { connect } from 'react-redux'
import PropTypes from 'prop-types'
import { withRouter } from 'react-router-dom'
import shallowCompare from 'react-addons-shallow-compare';

export class CategoriesHomepage extends Component {
	static propTypes = {
		requestCompleted: PropTypes.bool,
		search: PropTypes.arrayOf(
			PropTypes.object,
			PropTypes.string,
			PropTypes.number,
			PropTypes.arrayOf(PropTypes.string, PropTypes.number)
		)
	}

	static defaultProps = {
		requestCompleted: false,
		search: null
	}

	constructor (props) {
		super(props)
		this.handleSearchResult = this.handleSearchResult.bind(this)
	}

	shouldComponentUpdate(nextProps, nextState) {
		return shallowCompare(this, nextProps, nextState);
	}

	handleSearchResult () {
		return true
	}

	render () {
		let { search, requestCompleted } = this.props;
		let arrayMap;

		if (requestCompleted && Array.isArray(search) && search.length > 0) {
			search.splice(2, 0, '') // add empty element so that search form can be placed there

			arrayMap = search.map((element, index) => {
				Object.assign(element, {
					thumbnailUrl:
						window.location.protocol +
						'//' +
						window.location.hostname +
						(window.location.port.length === 0 ? '/' : ':8443/') +
						`thumbnails/thumbnail-${element._id}.png`
				})

				const key = index.toString();

				// only add search box on home page
				return (
					<div
						key={`${element}${key}`}
					>
						{index !== 2 ? (
							<PageUnit
								id={element._id}
								shortSlug={element.shortSlug}
								slug={element.slug}
								subtitle={element.subTitle}
								thumbnailUrl={element.thumbnailUrl}
								title={element.title}
							/>
						) : (
							<HomePageSearchForm onSearchInput={this.handleSearchResult} />
						)}
					</div>
				)
			})
		}

		const $thereAreNo = 'There are no entries for this term'

		if (requestCompleted && Array.isArray(search) && search.length === 0) {
			arrayMap = <p className='error'>{$thereAreNo}</p> // eslint-disable-line react/jsx-one-expression-per-line
		}

		const href = window.location.href
		const $title =
			href.indexOf('/category/') === -1
				? 'All categories'
				: href.substr(href.lastIndexOf('/') + 1)

		return (
			<main styleName='categories'>
				<h3>
					{$title}
				</h3>
				<section>
					{arrayMap}
				</section>
			</main>
		)
	}
}

export default connect()(withRouter(CategoriesHomepage))
