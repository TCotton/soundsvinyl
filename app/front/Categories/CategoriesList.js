/* eslint-disable */
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import PageUnit from '../Page/PageUnit'
import HomePageSearchForm from './HomePageSearchForm'

// import ErrorBoundary from '../errorBoundaries/ErrorBoundary';

export class CategoriesList extends Component {

	static propTypes = {
		current: PropTypes.number.isRequired,
		dataArray: PropTypes.arrayOf(
			PropTypes.oneOfType( [
				PropTypes.string,
				PropTypes.number,
				PropTypes.object
			] )
		).isRequired,
		pageSize: PropTypes.number.isRequired
	};

	constructor ( props ) {
		super( props );
	}

	render () {

		const { pageSize, current, dataArray } = this.props;
		const slicedArray = [ ...dataArray.slice( ( current - 1 ) * pageSize, ( pageSize * current ) ) ];
		slicedArray.splice( 2, 0, '' ); // add empty element so that search form can be placed there

		const arrayMap = slicedArray.map( ( element, index ) => {

			Object.assign( element, {
				thumbnailUrl: window.location.protocol + '//' + window.location.hostname + ( window.location.port.length === 0 ? '/' : ':8443/' ) + `thumbnails/thumbnail-${element._id}.png`
			} );

			const key = index.toString();

			return (
				<div key={`${element}${key}`}>
					{
						index !== 2 ?
							<PageUnit
								id={element._id}
								shortSlug={element.shortSlug}
								slug={element.slug}
								subtitle={element.subTitle}
								thumbnailUrl={element.thumbnailUrl}
								title={element.title}
							/> : <HomePageSearchForm onSearchInput={this.handleSearchResult} />
					}
				</div>
			)
		});

		return (
			<React.Fragment />
		)
	}
}

export default CategoriesList;
