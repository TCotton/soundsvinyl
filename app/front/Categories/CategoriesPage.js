import React, { Component } from 'react';
import './categoriesHomepage.scss';
import PageUnit from '../Page/PageUnit';
import HomePageSearchForm from './HomePageSearchForm';
import { connect } from 'react-redux';
import PropTypes from 'prop-types'

export class CategoriesPage extends Component {

	static propTypes = {
		requestCompleted: PropTypes.bool,
		search: PropTypes.arrayOf( PropTypes.object, PropTypes.string, PropTypes.number, PropTypes.arrayOf( PropTypes.string, PropTypes.number ) ).isRequired
	}

	static defaultProps = {
		requestCompleted: false
	}

	constructor (props) {
		super(props);
		this.handleSearchResult = this.handleSearchResult.bind( this );
	}

	handleSearchResult () {
		return true;
	}

	render () {
		const { search, requestCompleted } = this.props;
		let arrayMap;
		let $title = 'All categories';

		if (requestCompleted) {
			search.splice(2, 0, ''); // add empty element so that search form can be placed there
			arrayMap = search.map((element, index) => {

				Object.assign(element, {
					thumbnailUrl: window.location.protocol + '//' + window.location.hostname + (window.location.port.length === 0 ? '/' : ':8443/') + `thumbnails/thumbnail-${element._id}.png`
				});

				const key = index.toString();
				const searchForm = <HomePageSearchForm onSearchInput={this.handleSearchResult}  />;

				return (
					<div key={`${element}${key}`}>
						{
							index !== 2 ?
								<PageUnit
									id={element._id}
									shortSlug={element.shortSlug}
									slug={element.slug}
									subtitle={element.subTitle}
									thumbnailUrl={element.thumbnailUrl}
									title={element.title}
								/> : searchForm
						}
					</div>
				)
			});
		}

		return (
			<main styleName='categories'>
				<h3>
					{$title}
				</h3>
				<section>
					{arrayMap}
				</section>
			</main>
		)
	}
}

const mapStateToProps = ( state ) => {
	// redux needs refactoring
	if( state.search.length > 0 ) {
		return {
			search: state.search
		}
	}
}

export default connect(mapStateToProps)( CategoriesPage );
