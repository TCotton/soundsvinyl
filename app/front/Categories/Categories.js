import React, { Component } from 'react'
import axios from 'axios'
import { homeURI } from '../../helper_constants'
import CategoriesHomepage from './CategoriesHomepage'
import ErrorBoundary from '../errorBoundaries/ErrorBoundary'
import PropTypes from 'prop-types'
import { withRouter } from 'react-router-dom';

export class Categories extends Component {
	static propTypes = {
		category: PropTypes.string,
		history: PropTypes.shape({
			push: PropTypes.func
		}),
		location: PropTypes.shape({
				pathname: PropTypes.string,
				search: PropTypes.string
		}),
		match: PropTypes.shape({
			params: PropTypes.shape({
				id: PropTypes.string,
				tag: PropTypes.string
			})
		}),
	}

	static defaultProps = {
		category: undefined,
		history: {
			push: () => {},
		},
		location: {
			params: {
				pathname: '',
				search: ''
			}
		},
		match: {
			params: {
				id: ''
			}
		}
	}

	constructor (props) {
		super(props)
		this.state = this.getInitialState();
	}

	getInitialState () {

		return {
			error: false,
			loading: false,
			docs: [],
			requestCompleted: false,
		}
	}

	componentDidMount () {
		this.getRequestCall()
	}

	componentDidUpdate (prevProps, prevState) {
		const {
			match: {
				params: { tag }
			}
		} = this.props

		if (tag !== prevProps.match.params.tag) {
			this.getRequestCall()
		}

		let x = 0;
		let y = 0;

		Object.entries(this.props).forEach(([key, val]) =>
    	prevProps[key] !== val && console.log(`Prop '${key}' changed %s`, x++)
  	);
  	if (this.state) {
   	 Object.entries(this.state).forEach(([key, val]) =>
      prevState[key] !== val && console.log(`State '${key}' changed %s`, y++)
    );
  }
	}

	getRequestCall () {
		const { category } = this.props;

		// refactor both these API request into one request
		if (!category) {
			axios
				.get(
					`${homeURI}/apiV1/page/findAll`
				)
				.then(res => {
					this.setState({
						docs: res.data,
						requestCompleted: true
					})
				})
				.catch(error => {
					new Error(error.toString())
				})
		}

		if (category) {
			axios
				.get(`${homeURI}/apiV1/page/findbytag/${category}`)
				.then(res => {
					this.setState({
						docs: this.addNumberToDataArray(res.data),
						requestCompleted: true,
					})
				})
				.catch(error => {
					new Error(error.toString())
				})
		}
	}

/* 	handleOnChange (direction) {
		const paginationFunc = () => {
			const { page, total } = this.state;
			const maximum = Math.ceil(total / articlesPerPage);

			if (direction === 'left') {
				// current should be a minimum of one
				return page - 1 >= 1 ? page - 1 : 1;
			}
			if (direction === 'right') {
				// current should be maximum of Math.ceil(total / articlesPerPage)
				return page + 1 < maximum ? page + 1 : maximum;
			}
		}

		this.setState(
			{
				page: paginationFunc()
			},
			() => {
				return this.getRequestCall()
			}
		)
	} */

	addNumberToDataArray ([...data]) {
		data.forEach((currentValue, index) => {
			data[Number.parseInt(index)].position = Number.parseInt(index) + 1
		})

		return data;
	}

	render () {
		const { docs, requestCompleted } = this.state;
		console.dir(this.state);

		const { category } = this.props; // refactor -> use redux

		return (
			<ErrorBoundary>
				<CategoriesHomepage
					category={category}
					requestCompleted={requestCompleted}
					search={docs}
				/>
			</ErrorBoundary>
		)
	}
}

export default withRouter(Categories);
