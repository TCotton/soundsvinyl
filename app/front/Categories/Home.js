import React, { Component } from 'react';
import axios from 'axios';
import { homeURI } from '../../helper_constants';
import CategoriesHomepage from './CategoriesHomepage';
import ErrorBoundary from '../errorBoundaries/ErrorBoundary';

export class Home extends Component {

	constructor (props) {
		super(props);

		this.state = {
			pages: [],
			requestCompleted: false
		};
	}

	componentDidMount () {

		axios.get( `${homeURI}/apiV1/page/get` )
			.then( res => {

				this.setState({
					pages: res.data,
					requestCompleted: true
				})
			});
	}

	render () {
		const { pages, requestCompleted } = this.state;

		return (
			<ErrorBoundary>
				<CategoriesHomepage
					requestCompleted={requestCompleted}
					search={pages}
				/>
			</ErrorBoundary>
		)
	}
}

export default Home;
