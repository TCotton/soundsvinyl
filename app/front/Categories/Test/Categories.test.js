/* eslint-disable jest/no-disabled-tests */
import React from 'react'
import Categories from '../Categories'
import renderer from 'react-test-renderer';
import { MemoryRouter as Router } from 'react-router-dom'

describe('Component', () => {
	let component

	beforeEach(() => {
		component = renderer.create(
			<Router>
				<Categories />
			</Router>
			)
	})

	it('should be defined', () => {
		expect(component).toBeDefined()
	})

	it('should render correctly', function(){
		const component = renderer.create(
			<Router>
				<Categories />
			</Router>
		);
		const tree = component.toJSON();
		expect(tree).toMatchSnapshot();
	});
})
