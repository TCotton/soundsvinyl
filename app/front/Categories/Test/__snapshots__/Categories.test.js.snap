// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component should render correctly 1`] = `
Array [
  <h2>
    Something went wrong.
  </h2>,
  <details
    className="ErrorBoundary__fragmentDetails___2-AcC"
  >
    Invariant Violation: Could not find "store" in the context of "Connect(withRouter(CategoriesHomepage))". Either wrap the root component in a &lt;Provider&gt;, or pass a custom React context provider to &lt;Provider&gt; and the corresponding React context consumer to Connect(withRouter(CategoriesHomepage)) in connect options.
    <br />
    
    in ConnectFunction (created by Categories)
    in ErrorBoundary (created by Categories)
    in Categories (created by Context.Consumer)
    in withRouter(Categories)
    in Router (created by MemoryRouter)
    in MemoryRouter
  </details>,
]
`;
